dao/src/main/java/org/thingsboard/server/dao/entity/BaseEntityService.java

```diff
@@ -79,7 +85,53 @@ public PageData<EntityData> findEntityDataByQuery(TenantId tenantId, CustomerId
         validateId(tenantId, INCORRECT_TENANT_ID + tenantId);
         validateId(customerId, INCORRECT_CUSTOMER_ID + customerId);
         validateEntityDataQuery(query);
-        return this.entityQueryDao.findEntityDataByQuery(tenantId, customerId, query);
+
+        if (EntityFilterType.RELATIONS_QUERY.equals(query.getEntityFilter().getType())
+                || EntityFilterType.SINGLE_ENTITY.equals(query.getEntityFilter().getType())
+                || StringUtils.isNotEmpty(query.getPageLink().getTextSearch())) {
+            return this.entityQueryDao.findEntityDataByQuery(tenantId, customerId, query);
+        }
+
+        // 1 step - find entity data by filter and sort columns
+        PageData<EntityData> entityDataByQuery = findEntityIdsByFilterAndSorterColumns(tenantId, customerId, query);
+        if (entityDataByQuery == null || entityDataByQuery.getData().isEmpty()) {
+            return entityDataByQuery;
+        }
+        // 2 step - find entity data by entity ids from the 1st step
+        PageData<EntityData> result = findEntityDataByEntityIds(tenantId, customerId, query, entityDataByQuery.getData());
+        return new PageData<>(result.getData(), entityDataByQuery.getTotalPages(), entityDataByQuery.getTotalElements(), entityDataByQuery.hasNext());
+    }
+
+    private PageData<EntityData> findEntityIdsByFilterAndSorterColumns(TenantId tenantId, CustomerId customerId, EntityDataQuery query) {
+        List<EntityKey> entityFields = null;
+        List<EntityKey> latestValues = null;
+        if (query.getPageLink().getSortOrder() != null) {
+            if (query.getEntityFields() != null) {
+                entityFields = query.getEntityFields().stream()
+                        .filter(entityKey -> entityKey.getKey().equals(query.getPageLink().getSortOrder().getKey().getKey()))
```

Comment:
Filter fields that are sort columns and fields used in ```query.getKeyFilters()```, since they are already fetched for filtering.